File: II-159-2/Sets.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
#define all(v) (v).begin(), (v).end()
struct Sets
{
	vi operate( vi A, vi B, string operation )
	{
		vi result;
		sort( all(A) ); sort( all(B) );
		if ( operation == "INTERSECTION" )
			set_intersection( all(A), all(B), back_inserter( result ) );
		else if ( operation == "UNION" )
			set_union( all(A), all(B), back_inserter( result ) );
		else
		{
			vi C, D;
			set_intersection( all(A), all(B), back_inserter( C ) );
			set_union( all(A), all(B), back_inserter( D ) );
			set_difference( all(D), all(C), back_inserter( result ) );
		}
		return result;
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vi> A0; vector<vi> A1; vector<string> A2; vector<vi> A3;
	{ /*Test 0*/int _A[] = 
	{1,2,3,4}
	; vi A(_A, _A + (sizeof(_A) / sizeof(_A[0]))); A0.push_back( A ); int _B[] = 
	{3,4,5,6}
	; vi B(_B, _B + (sizeof(_B) / sizeof(_B[0]))); A1.push_back( B ); string operation = 
	"INTERSECTION"
	; A2.push_back( operation ); int _result[] = 
	{ 3,  4 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A3.push_back( result );  }
	{ /*Test 1*/int _A[] = 
	{1,2,3,4}
	; vi A(_A, _A + (sizeof(_A) / sizeof(_A[0]))); A0.push_back( A ); int _B[] = 
	{3,4,5,6}
	; vi B(_B, _B + (sizeof(_B) / sizeof(_B[0]))); A1.push_back( B ); string operation = 
	"UNION"
	; A2.push_back( operation ); int _result[] = 
	{ 1,  2,  3,  4,  5,  6 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A3.push_back( result );  }
	{ /*Test 2*/int _A[] = 
	{432,756,123}
	; vi A(_A, _A + (sizeof(_A) / sizeof(_A[0]))); A0.push_back( A ); int _B[] = 
	{534,76,1209}
	; vi B(_B, _B + (sizeof(_B) / sizeof(_B[0]))); A1.push_back( B ); string operation = 
	"INTERSECTION"
	; A2.push_back( operation ); vi result; A3.push_back( result );  }
	{ /*Test 3*/int _A[] = 
	{6,5,7,4}
	; vi A(_A, _A + (sizeof(_A) / sizeof(_A[0]))); A0.push_back( A ); int _B[] = 
	{7,6,4,10}
	; vi B(_B, _B + (sizeof(_B) / sizeof(_B[0]))); A1.push_back( B ); string operation = 
	"SYMMETRIC DIFFERENCE"
	; A2.push_back( operation ); int _result[] = 
	{ 5,  10 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A3.push_back( result );  }
	{ /*Test 4*/int _A[] = 
	{342,654,897,312,76,23,78}
	; vi A(_A, _A + (sizeof(_A) / sizeof(_A[0]))); A0.push_back( A ); int _B[] = 
	{21,43,87,98,23,756,897,234,645,876,123}
	; vi B(_B, _B + (sizeof(_B) / sizeof(_B[0]))); A1.push_back( B ); string operation = 
	"SYMMETRIC DIFFERENCE"
	; A2.push_back( operation ); int _result[] = 
	{ 21,  43,  76,  78,  87,  98,  123,  234,  312,  342,  645,  654,  756,  876 }
	; vi result(_result, _result + (sizeof(_result) / sizeof(_result[0]))); A3.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A3[i], operate( A0[i], A1[i], A2[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	Sets ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE

File: II-159-2/Sets.cs
using System;
using System.Collections;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
 
public class Sets
{
	public int[] operate( int[] A, int[] B, string operation )
	{
		ArrayList a = new ArrayList();
		ArrayList b = new ArrayList();
		ArrayList r = new ArrayList();
		for( int i = 0; i < A.Length; i++ )
		{
			if ( !a.Contains( A[i] ) ) a.Add( A[i] );
		}
		for( int i = 0; i < B.Length; i++ )
		{
			if ( !b.Contains( B[i] ) ) b.Add( B[i] );
		}
		switch( operation )
		{
			case "INTERSECTION":
				for( int i = 0; i < a.Count; i++ )
				{
					if ( !r.Contains( a[i] ) && b.Contains( a[i] ) ) r.Add( a[i] );
				}
				break;
			case "UNION":
				for( int i = 0; i < a.Count; i++ )
				{
					if ( !r.Contains( a[i] ) ) r.Add( a[i] );
				}
				for( int i = 0; i < b.Count; i++ )
				{
					if ( !r.Contains( b[i] ) ) r.Add( b[i] );
				}
				break;
			case "SYMMETRIC DIFFERENCE":
				for( int i = 0; i < a.Count; i++ )
				{
					if ( !r.Contains( a[i] ) && !b.Contains( a[i] ) ) r.Add( a[i] );
				}
				for( int i = 0; i < b.Count; i++ )
				{
					if ( !r.Contains( b[i] ) && !a.Contains( b[i] )  ) r.Add( b[i] );
				}
				break;
		}
		int[] result = new int[r.Count];
		for( int i = 0; i < r.Count; i++ )
		{
			result[i] = (int) r[i];
		}
		Array.Sort( result );
		return result;
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new Sets()).operate(new int[] {1,2,3,4}, new int[] {3,4,5,6}, "INTERSECTION"),new int[] { 3,  4 });
		eq(2,(new Sets()).operate(new int[] {1,2,3,4}, new int[] {3,4,5,6}, "UNION"),new int[] { 1,  2,  3,  4,  5,  6 });
		eq(3,(new Sets()).operate(new int[] {432,756,123}, new int[] {534,76,1209}, "INTERSECTION"),new int[] { });
		eq(4,(new Sets()).operate(new int[] {6,5,7,4}, new int[] {7,6,4,10}, "SYMMETRIC DIFFERENCE"),new int[] { 5,  10 });
		eq(5,(new Sets()).operate(new int[] {342,654,897,312,76,23,78}, new int[] {21,43,87,98,23,756,897,234,645,876,123}, "SYMMETRIC DIFFERENCE"),new int[] { 21,  43,  76,  78,  87,  98,  123,  234,  312,  342,  645,  654,  756,  876 });
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( t.TotalMilliseconds / 1000 );
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, double a, double b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, double[] a, double[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(double[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}
// END CUT HERE
}
